[Speculative Decoding] EAGLE Implementation with Top-1 proposer

abhigoyal1997 opened the pull request:
This PR adds support for the EAGLE draft model.

Fix SafeAILab/EAGLE#43

abhigoyal1997 committed:
initial changes to support EAGLE

abhigoyal1997 committed:
handling hidden_states in case of bonus tokens since EAGLE will need it

abhigoyal1997 committed:
enabling CUDA graph

abhigoyal1997 committed:
adding E2E test and formatting

abhigoyal1997 commented:
/ready

abhigoyal1997 committed:
minor bug fix in graph capture

abhigoyal1997 committed:
fixing broadcasting of hidden states in distributed worker

abhigoyal1997 committed:
formatting

abhigoyal1997 committed:
Merge branch 'main' of github.fkinternal.com:abhinav-goyal/vllm into eagle

abhigoyal1997 committed:
formatting

caddfa31434 commented on vllm/model_executor/models/eagle.py:
```diff
@@ -0,0 +1,84 @@
+        intermediate_tensors: Optional[IntermediateTensors] = None,
+    ) -> torch.Tensor:
+
+        tok_embeds = self.model.model.embed_tokens(input_ids)
```
Should we mask fill token embeddings where 'input_positions' equals zero?

abhigoyal1997 replied:
Right now, the hidden states are masked for input position 0. We can mask token embeddings as well, but it's unlikely that it'll affect the results, right? Or do you see this affecting anything?

caddfa31434 replied:
The inputs_embeds after fc will have some subtle differences compared to the original EAGLE

caddfa31434 commented:
For some reason, Eagle seems to have removed the input_layernorm ： https://github.com/SafeAILab/EAGLE/blob/main/eagle/model/cnets.py#L419

abhigoyal1997 commented:
> For some reason, Eagle seems to have removed the input_layernorm ： https://github.com/SafeAILab/EAGLE/blob/main/eagle/model/cnets.py#L419

Yes, I saw that recently. But making that change would mean either changing the decoder layer to have input layernorm as optional or rewriting the decoder layer just for EAGLE. Both these options would reduce the freedom to use any decoder with EAGLE.

abhigoyal1997 commented on tests/spec_decode/e2e/test_eagle_correctness.py:
```diff
@@ -0,0 +1,224 @@
+    * Test greedy equality under various number of speculative tokens.
+
+With those tests, we can say at least, Medusa would not break the
+correctess for the target model outputs.
```
Typo: to be fixed with next commit.

cadedaniel commented:
cc @comaniac @sroy745

abhigoyal1997 committed:
Merge branch 'main' of github.fkinternal.com:abhinav-goyal/vllm into eagle

abhigoyal1997 committed:
Masking position=0 in inputs for EAGLE

abhigoyal1997 committed:
reformatting

abhigoyal1997 committed:
Fixing the order of execution for scorer and proposer in non-driver worker

sroy745 commented on vllm/spec_decode/multi_step_worker.py:
```diff
@@ -153,6 +153,12 @@ def _expand_execute_model_request(
+
+        if isinstance(updated_execute_model_req.previous_hidden_states,
+                      HiddenStates):
+            updated_execute_model_req.previous_hidden_states\
```
In this method we are changing the order of sequences in updated_execute_model_req compared to the original execute_model_req depending on whether sequences have bonus tokens or not. Wondering if order of the hidden states need to be aligned with the new ordering in updated_execute_model_req and if it is being handled in expand_with_bonus_tokens? E.g. in expand_with_bonus_tokens I see we are adding the hidden states for the bonus tokens before the one without the bonus token which is different compared to the order of seq_group_metadata in updated_execute_model_req. Is that expected?

abhigoyal1997 replied:
Actually, in expand_with_bonus_tokens, the hidden states for tokens before the bonus are added first since bonus_token_previous_hidden_states contains the hidden states of the token before the bonus. Maybe I should rename the variable name to avoid confusion.

abhigoyal1997 replied:
You are right, in this particular case, the order will not match. But HiddenStates init has an assertion that forces all seq groups to have only one sequence. So, I wrote the code with that assumption in mind. However I can change this, it shouldn't be a big change.

Yes, I can add the test as well but it should be ok to test with this assertion in mind, right? Or should we force-test the case where seq_group can have multiple sequences?

sroy745 replied:
Thanks for adding the test. Is there any existing test which would validate the assertion that each seq_group has one sequence. If not I was wondering if you could add a small test in https://github.com/vllm-project/vllm/blob/main/tests/test_sequence.py to validate this. I don't think the current test that you added would fail if we were to remove that assertion.

abhigoyal1997 replied:
I thought this was a broader assumption as we don't have support for beam search in speculative decoding. In that case, do you mean we should add a test to ensure that in speculative decoding each seq_group has one sequence?

sroy745 replied:
Sorry for the confusion. No need for this test.

sroy745 commented on vllm/model_executor/models/eagle.py:
```diff
@@ -0,0 +1,86 @@
+    def sampler(self):
+        return self.model.sampler
+
+    def forward(
```
Wondering if you have been able to run any comparison between eagle and medusa and how the results compare?

sroy745 replied:
Reopening this comment. Wondering if you have been able to compare the performance of the eagle model with that of medusa?

abhigoyal1997 replied:
Sorry, I thought I already replied to this.

Yes, I compared the two for Meta-llama-3-8b but found that the overhead for speculation is currently very high for EAGLE. Because of this, the performance of EAGLE (from https://huggingface.co/yuhuili/EAGLE-LLaMA3-Instruct-8B) was similar to that of without any speculation even though the average tokens/step was around 1.25 on MT-Bench.

I am also trying to run a more comprehensive benchmarking or no spec vs Medusa vs EAGLE and will share the results when complete.

sroy745 reviewed (commented):
Thanks for the PR. Left some comments. PTAL

abhigoyal1997 commented on vllm/worker/worker_base.py:
```diff
@@ -239,10 +239,26 @@ def execute_model(
                     execute_model_req.finished_requests_ids))
             num_steps = execute_model_req.num_steps
 
+            if isinstance(execute_model_req.previous_hidden_states,
```
Yes. I'll make the change.

abhigoyal1997 committed:
Adding hidden state propagation to _execute_model_spmd

abhigoyal1997 committed:
Adding CUDA graph tests for medusa and eagle. Renaming mlp to medusa or eagle where required.

abhigoyal1997 commented on vllm/spec_decode/spec_decode_worker.py:
```diff
@@ -496,15 +507,21 @@ def _run_non_driver_rank(self) -> bool:
-        # delegate how many times it runs to the proposer.
-        for _ in range(max(num_lookahead_slots, 1)):
-            self.proposer_worker.execute_model()
+        if data["no_spec"]:
```
Yes. I'll add a comment in the code.

abhigoyal1997 commented on vllm/worker/model_runner.py:
```diff
@@ -1387,6 +1402,12 @@ def execute_model(
             if model_input.is_prompt:
                 hidden_states = hidden_or_intermediate_states.index_select(
                     0, indices)
+                prefill_hidden_states = hidden_or_intermediate_states.roll(
```
Yes, that should be fine.

abhigoyal1997 committed:
Moving hidden states shift to spec_decode_worker

abhigoyal1997 committed:
formatting

abhigoyal1997 committed:
Merge branch 'vllm-project:main' into eagle

abhigoyal1997 commented:
@sroy745 Thanks for the review. I've made the changes and responded to your comments. PTAL

sroy745 commented on vllm/sequence.py:
```diff
@@ -939,18 +943,32 @@ class HiddenStates:
+            self,
+            seq_group_metadata_list: List[SequenceGroupMetadata],
+            hidden_states: torch.Tensor,
+            bonus_token_previous_hidden_states: Optional[torch.Tensor] = None):
```
nit: Can we rename this to last_non_bonus_token_hidden_states?

sroy745 commented on vllm/spec_decode/spec_decode_worker.py:
```diff
@@ -632,10 +661,12 @@ def _verify_tokens(
             accepted_index = accepted_token_ids + 1  # Convert -1 to 0
             accepted_index = accepted_index.count_nonzero(dim=1).add_(-1)
             index = accepted_index[:, None, None].expand(-1, 1, hs_size)
+            bonus_token_previous_hidden_states = hidden_states[:, -2]  # b x d
```
nit : Wondering if we can rename bonus_token_previous_hidden_states to last_non_bonus_token_hidden_states?

sroy745 reviewed (commented):
Thanks for addressing my comments. Left a few more comments (except one all other are nits). PTAL.

abhigoyal1997 committed:
Adding vocab truncation to EAGLE

abhigoyal1997 commented on vllm/sequence.py:
```diff
@@ -939,18 +943,32 @@ class HiddenStates:
+        # Adding dummy hidden_states to this to maintain same shape
+        self.last_non_bonus_token_hidden_states = torch.cat([
+            self.hidden_states,
+            hidden_states if bonus_token_previous_hidden_states is None else
```
Yes, that would also be fine. This is never to be used anyways.

abhigoyal1997 committed:
Minor changes and fixes. Adding expand model request and hidden states sync test

abhigoyal1997 committed:
Merge branch 'vllm-project:main' into eagle

abhigoyal1997 commented on tests/spec_decode/test_multi_step_worker.py:
```diff
@@ -690,3 +691,55 @@ def test_use_draft_model_runner_advance_step():
+
+
+@torch.inference_mode()
+def test_expand_execute_model_request_sync_with_expand_hidden_states():
```
Currently, this assumes `HiddenStates` are only used when all seq_groups have 1 sequence each (as per the assertion in `HiddenStates.__init__`). If that assertion needs to be relaxed, we can modify this test to add a more general case.

The way this works is by matching the order of seqs in expanded seq_group_metadata_list with the order of previous_hidden_states in expanded execute_model_request.

cadedaniel replied:
Nice. What happens if a request is aborted?

cadedaniel replied:
OK I think there is an issue here where a sequence may be "paused" then "resumed" later. so this should only remove things that are confirmed aborted.

but we can fix that later.

cadedaniel replied:
can you add a comment for this?

abhigoyal1997 replied:
Added this comment to `HiddenStates.prune`

abhigoyal1997 commented on vllm/model_executor/models/eagle.py:
```diff
@@ -0,0 +1,86 @@
+                            config.model.hidden_size,
+                            bias=False)
+
+        self.token_map = None
```
This is used for vocab truncation (as introduced in Medusa as well). The code wasn't complete for this, I've made the changes necessary to add complete support and have added the comments in code as well.

cadedaniel replied:
Can this be disabled? If not, can you point me to the logic in the reference implementation?

abhigoyal1997 replied:
This is disabled by default and is only used if the checkpoint file of the draft model has an explicit token_map tensor and config has an optional attribute truncated_vocab_size < vocab_size. I am adding this clarification to both EAGLE and Medusa in the comments and the docstring.

This wasn't there in any of the reference implementations but we have found it to work well as it reduces overhead with minimal impact on acceptance rate. To use it, you must find the most frequent tokens in your target data distribution and add them as token_map to the draft checkpoint.

cadedaniel replied:
It would be awesome to have this explanation in the docs or in a docstring. Cool technique.

Seems also the testing story is weak for this (correct me if wrong), but we can follow up later since it's not new to this PR.

abhigoyal1997 replied:
Added the explanation to both Medusa's and EAGLE's doctring.

richardliaw commented:
Thanks for the contribution! Do we have any benchmarks on this PR?

abhigoyal1997 committed:
Merge branch 'main' into eagle

abhigoyal1997 commented:
> Thanks for the contribution! Do we have any benchmarks on this PR?

Hi. We don't have the comprehensive benchmarks yet, but from some tests on Llama 3 8b this is not giving any gains as of now (using https://huggingface.co/yuhuili/EAGLE-LLaMA3-Instruct-8B) because of high overhead of EAGLE (mostly due to how auto regressive spec. decode works as of now). With an internal model I did observe significant gains (1.3x faster for 8B model) due to higher acceptance rate. But I think it will improve as other optimizations of speculative decoding are completed.

abhigoyal1997 committed:
Merge branch 'vllm-project:main' into eagle

abhigoyal1997 committed:
Merge branch 'vllm-project:main' into eagle

abhigoyal1997 commented:
@sroy745 I've addressed the new comments you had added. PTAL.

caddfa31434 commented:
@abhigoyal1997  Thanks for the great PR!  I'm having trouble figuring out how to update the prefill_hidden_states at each step within the MultiStepWorker. Could you please provide some guidance?

caddfa31434 commented:
@abhigoyal1997  I am under the impression that in the Eagle proposer, each step necessitates the hidden states from the previous step as input. If I am mistaken, please correct me.

abhigoyal1997 commented:
> @abhigoyal1997 I am under the impression that in the Eagle proposer, each step necessitates the hidden states from the previous step as input. If I am mistaken, please correct me.

Yes, it does but prefill_hidden_states are only used for prefill step. For decode step, the hidden states are propagated within draft_model_runner (https://github.com/flipkart-incubator/vllm/blob/3c24f4bbbe7c46ec2354fdcd00e9f1d53d3dfaad/vllm/spec_decode/draft_model_runner.py#L339C13-L339C26).

caddfa31434 commented:
> > @abhigoyal1997 I am under the impression that in the Eagle proposer, each step necessitates the hidden states from the previous step as input. If I am mistaken, please correct me.
> 
> Yes, it does but prefill_hidden_states are only used for prefill step. For decode step, the hidden states are propagated within draft_model_runner (https://github.com/flipkart-incubator/vllm/blob/3c24f4bbbe7c46ec2354fdcd00e9f1d53d3dfaad/vllm/spec_decode/draft_model_runner.py#L339C13-L339C26).

Please ignore the 'prefill_hidden_states'. What I mean is, at each step of the 'multi_step_worker', update the 'execute_model_req.previous_hidden_states' with 'model_output.return_hidden_states'(https://github.com/vllm-project/vllm/blob/3c24f4bbbe7c46ec2354fdcd00e9f1d53d3dfaad/vllm/spec_decode/multi_step_worker.py#L88).

abhigoyal1997 commented:
> > > @abhigoyal1997 I am under the impression that in the Eagle proposer, each step necessitates the hidden states from the previous step as input. If I am mistaken, please correct me.
> > 
> > 
> > Yes, it does but prefill_hidden_states are only used for prefill step. For decode step, the hidden states are propagated within draft_model_runner (https://github.com/flipkart-incubator/vllm/blob/3c24f4bbbe7c46ec2354fdcd00e9f1d53d3dfaad/vllm/spec_decode/draft_model_runner.py#L339C13-L339C26).
> 
> Please ignore the 'prefill_hidden_states'. What I mean is, at each step of the 'multi_step_worker', update the 'execute_model_req.previous_hidden_states' with 'model_output.return_hidden_states'(
> 
> https://github.com/vllm-project/vllm/blob/3c24f4bbbe7c46ec2354fdcd00e9f1d53d3dfaad/vllm/spec_decode/multi_step_worker.py#L88
> 
> ).

Actually, EAGLE only supports TP1DraftModelRunner in this PR and model_runner also doesn't support hidden_states as inputs. Therefore, this is not needed unless we need to support TP > 1 for EAGLE.

abhigoyal1997 committed:
Merge branch 'main' into eagle

caddfa31434 commented:
@abhigoyal1997  In the current setup, if I set num_speculative_tokens to 5, does it means that the proposal will indeed supply 5 tokens as input for generation, all using the exact same previous_hidden_states. If there's any critical detail I've overlooked, please let me know.

abhigoyal1997 commented:
> @abhigoyal1997 In the current setup, if I set num_speculative_tokens to 5, does it means that the proposal will indeed supply 5 tokens as input for generation, all using the exact same previous_hidden_states. If there's any critical detail I've overlooked, please let me know.

EAGLE will propose 5 tokens, but not the same state. Please check execute_model in draft_model_runner.py (line 315 onwards). The hidden_states output by the proposer at L339 is provided as input to the next step via kwargs (L335). As for the 1st step, lines 315-229 prepare the previous_hidden_states output by scorer and the same is again added to kwargs at L335 (since it is stored in hidden_states).

sroy745 reviewed (commented):
Thanks for addressing the comments. I've left a few minor comments, including a request for an additional test.

Otherwise, the PR looks good to me. Could you also share your plans regarding the performance issues you mentioned earlier https://github.com/vllm-project/vllm/pull/6830#discussion_r1701358711

cc: @cadedaniel

abhigoyal1997 committed:
Removing commented code and a minor comment fix

abhigoyal1997 committed:
formatting

abhigoyal1997 commented:
> Thanks for addressing the comments. I've left a few minor comments, including a request for an additional test.
> 
> Otherwise, the PR looks good to me. Could you also share your plans regarding the performance issues you mentioned earlier [#6830 (comment)](https://github.com/vllm-project/vllm/pull/6830#discussion_r1701358711)
> 
> cc: @cadedaniel

For the additional test, it's a bit unclear what exactly is required. Can you please elaborate?

About performance issues, I think they are not specific to EAGLE but spec. decode in general. EAGLE is nothing but a draft model that takes the hidden state of the target model as additional inputs. So, the autoregressive nature of the proposal is causing lower performance (compared to Medusa which is non-autoregressive) which should be fixed when the draft model proposal is optimized.

abhigoyal1997 committed:
Merge branch 'vllm-project:main' into eagle

jacob-crux commented on vllm/model_executor/models/eagle.py:
```diff
@@ -0,0 +1,140 @@
+            else:
+                model_weights[f"model.{name}"] = loaded_weight
+
+        lm_head_weight = model_weights.pop("lm_head.weight")
```
I'm testing with eagle llama3, but it seems there are still compatibility issues.
The `EAGLE-LLaMA3-Instruct-8B` model has an error because there is no lm_head in that part.
Could you share the test results for the llama3 8b model?

abhigoyal1997 replied:
Yes, it's because this model doesn't have lm_head. But since this was trained using the frozen lm_head from target model, we can add it manually from Llama3 checkpoint. I am uploading a version here: https://huggingface.co/abhigoyal/EAGLE-LLaMA3-Instruct-8B-vllm. Please try that. As I mentioned in another comment, this still doesn't give speed-up yet, but we have seen 1.3x seed-up in internal experiments.

sroy745 replied:
I was able to get the e2e tests (tests/spec_decode/e2e/test_eagle_correctness.py) to work with MAIN_MODEL = "meta-llama/Meta-Llama-3-8B" and SPEC_MODEL = "abhigoyal/EAGLE-LLaMA3-Instruct-8B-vllm". However I am wondering if it makes sense to add a comment regarding the incompatibility of yuhuili/EAGLE-LLaMA3-Instruct-8B with this implementation. Also if possible can you add a pointer to the example implementation of how you generated the working checkpoint (abhigoyal/EAGLE-LLaMA3-Instruct-8B-vllm) from the original one.

abhigoyal1997 replied:
Sure, I'll add the comments and explanation on how to create the working checkpoint.

jacob-crux replied:
![plot_ttft_decode_thru](https://github.com/user-attachments/assets/84aa0e85-1be0-438e-889a-8d7b536c2a30)
The above results are graphs showing Time To First Token (TTFT) and Decode Throughput.
I conducted a performance comparison experiment with a fixed number of concurrent users, running a total of five configurations(one without speculative decoding and four with `--num-speculative-tokens` set to 2, 3, 4, and 5).

Using a single A100 GPU with LLaMA3 8B, I observed that using `--num-speculative-tokens 5` resulted in a lower draft acceptance rate, making it difficult to achieve significant speed improvements.

abhigoyal1997 replied:
@jacob-crux Thanks for trying it out and sharing these results. These are very helpful. May I ask what dataset did you use for this?

jacob-crux replied:
@abhigoyal1997 
I used sharegpt dataset.

cadedaniel replied:
How does the draft acceptance rate look? A while ago I checked top1 acceptance rate for Eagle/llama2 and it was about .7 on sharegpt. This is considerably better than a 68m draft model which got .5-.6, and Medusa's top5 of .8.

sroy745 replied:
Ran the serving_tests in nightly-benchmarks with the config below and found the following acceptance rates. I used the target model as "meta-llama/Meta-Llama-3-8B-Instruct" and draft model as "abhigoyal/EAGLE-LLaMA3-Instruct-8B-vllm"

k=1 Draft acceptance rate: 0.637, System efficiency: 0.819

k=2 Draft acceptance rate: 0.445, System efficiency: 0.594

k=3  Draft acceptance rate: 0.377, System efficiency: 0.463

k=4  Draft acceptance rate: 0.335, System efficiency: 0.374

The acceptance rate is lower than what you observed especially at larger k.

```
[
    {
        "test_name": "serving_llama8B_tp4_sharegpt_specdecode",
        "qps_list": [1],
        "server_parameters": {
            "model": "meta-llama/Meta-Llama-3-8B-Instruct",
            "disable_log_requests": "", 
            "tensor_parallel_size": 4,
            "swap_space": 16, 
            "speculative_model": "abhigoyal/EAGLE-LLaMA3-Instruct-8B-vllm",
            "num_speculative_tokens": 3,
            "speculative_draft_tensor_parallel_size": 1,
            "use_v2_block_manager": "",
            "disable_custom_all_reduce": ""
        },
        "client_parameters": {
            "model": "meta-llama/Meta-Llama-3-8B-Instruct",
            "backend": "vllm",
            "dataset_name": "sharegpt",
            "dataset_path": "./ShareGPT_V3_unfiltered_cleaned_split.json",
            "num_prompts": 100 
        }
    }
]

```

abhigoyal1997 replied:
One difference is that the EAGLE official implementation for some reason has dropped the input_layernorm for 1st decoder layer. Since in decoder layers we have this is not the case, it might be the reason for different results.

cadedaniel replied:
maybe. whatever the issue is, it causes a big accuracy drop after k=1.

jacob-crux reviewed (commented):
It would be really useful if you could share the test case for the llama model.

abhigoyal1997 committed:
adding comments to clarify compatibility of eagle checkpoint in eagle.py

abhigoyal1997 committed:
Merge branch 'vllm-project:main' into eagle

abhigoyal1997 committed:
fixing model_cls resolution in eagle

abhigoyal1997 committed:
fixing model_cls resolution in eagle

abhigoyal1997 committed:
Merge branch 'main' into eagle

cadedaniel commented on tests/spec_decode/e2e/test_eagle_correctness.py:
```diff
@@ -0,0 +1,268 @@
+MAX_SPEC_TOKENS = 4
+
+# precision
+PRECISION = "float32"
```
Is this the recommended configuration for production deployments? We may be hiding a correctness issue that appears in fp16.

abhigoyal1997 replied:
For prod, we can use FP16 as well and it should work fine for all practical purposes. However, I've noticed that when using FP16, the distribution of the target model differs (though minimally). This sometimes results in different tokens being selected by the target model depending on it being used as a verifier or an independent generator.

For a more concrete example, please take a look at this: https://github.com/vllm-project/vllm/pull/4978#discussion_r1627603023

cadedaniel commented on tests/spec_decode/e2e/test_eagle_correctness.py:
```diff
@@ -0,0 +1,268 @@
+])
+@pytest.mark.parametrize("batch_size", [1, 32])
+@pytest.mark.parametrize("seed", [1])
+def test_eagle_e2e_greedy_correctness_cuda_graph(baseline_llm_generator,
```
Nice test, thanks for thinking of & adding this case

abhigoyal1997 replied:
Credits to @sroy745 for suggestting

cadedaniel commented on vllm/model_executor/models/eagle.py:
```diff
@@ -0,0 +1,143 @@
+from vllm.transformers_utils.configs.eagle import EAGLEConfig
+
+
+class EAGLE(nn.Module):
```
nit: docstring pointing to reference implementation / paper

cadedaniel commented on vllm/sequence.py:
```diff
@@ -1044,19 +1048,36 @@ class HiddenStates:
+            self,
+            seq_group_metadata_list: List[SequenceGroupMetadata],
+            hidden_states: torch.Tensor,
+            last_non_bonus_token_hidden_states: Optional[torch.Tensor] = None):
```
Add a docstring for the new field? seems non obvious

abhigoyal1997 replied:
Also renaming last_non_bonus_token_hidden_states to second_last_token_hidden_states as that is more appropriate.

cadedaniel commented on vllm/sequence.py:
```diff
@@ -1087,7 +1128,7 @@ class ExecuteModelRequest:
     running_queue_size: int = 0
     # Optional hidden states from prior step.
-    previous_hidden_states: Optional[HiddenStates] = None
+    previous_hidden_states: Union[HiddenStates, torch.Tensor, None] = None
```
nit: push this complexity to `HiddenStates` definition:
`HiddenStates = Union[ActualHiddenStates, torch.Tensor]`
...
`previous_hidden_states: Optional[HiddenStates] = None`

need to think of a better name for `ActualHiddenStates`. I am not sure why it would be a torch tensor sometimes and a normal class in other cases.

abhigoyal1997 replied:
During the decode stage, the HiddenStates class stores only the hidden states of the last tokens that are propagated across steps. For the prefill stage we need hidden states of all tokens and there's no propagation across steps, so the hidden states are a simple tensor as we don't need any of the features implemented by HiddenStates class.

If it helps, we can rename HiddenStates class to DecodeHiddenStates and use it as following:
```
PrefillHiddenStates = torch.Tensor
HiddenStates = Union[PrefillHiddenStates, DecodeHiddenStates]
```

Otherwise, we can store prefill hidden states also using the same HiddenStates class and make seq_group_metadata_list optional (basically if seq_group_metadata_list is optional, we use it as a simple wrapper on tensor).

cadedaniel replied:
The second solution you propose adds less complexity to vLLM, but the first works too. Will defer to you as it's a nit.

abhigoyal1997 replied:
States for the prefill step are now also handled by the HiddenStates class itself.

cadedaniel commented on vllm/worker/model_runner.py:
```diff
@@ -1205,6 +1206,15 @@ def capture_model(self, kv_caches: List[List[torch.Tensor]]) -> None:
         seq_lens = torch.ones(max_batch_size, dtype=torch.int32).cuda()
         block_tables = torch.from_numpy(self.graph_block_tables).cuda()
+        previous_hidden_states = None
+        if "previous_hidden_states" in inspect.signature(
+                self.model.forward).parameters:
```
Is there a better way to do this besides dynamic inspection ?

cadedaniel replied:
let's add a comment what this is for? e.g. EAGLE

cadedaniel replied:
OK. I'd feel better if it were in a separate helper function so it is hidden from view, but acknowledge that it's not much better..

cadedaniel commented on vllm/worker/worker_base.py:
```diff
@@ -385,3 +400,20 @@ def execute_method(self, method, *args, **kwargs):
             raise e
+
+
+def extract_previous_hidden_states(
```
add a docstring?

cadedaniel commented on vllm/worker/worker_base.py:
```diff
@@ -385,3 +400,20 @@ def execute_method(self, method, *args, **kwargs):
+        data: Union[ExecuteModelRequest, dict]) -> dict:
+    output = {}
+
+    if isinstance(data, dict):
+        if "previous_hidden_states" in data:
+            output["previous_hidden_states"] = data["previous_hidden_states"]
+    else:
+        if isinstance(data.previous_hidden_states, torch.Tensor):
+            output["previous_hidden_states"] = data.previous_hidden_states
+        elif isinstance(data.previous_hidden_states, HiddenStates):
+            output["previous_hidden_states"] = data.previous_hidden_states\
+                .hidden_states
```
Can you explain in a comment why we have such complexity?

abhigoyal1997 replied:
Will add a comment if needed after resolving https://github.com/vllm-project/vllm/pull/6830/files#r1714091225

cadedaniel replied:
yeah would be great if this could go away. the less things to unit test the better.

cadedaniel commented on vllm/spec_decode/spec_decode_worker.py:
```diff
@@ -360,6 +365,9 @@ def execute_model(
         # communication to inform them.
         broadcast_dict = dict(
             num_lookahead_slots=num_lookahead_slots,
+            no_spec=num_lookahead_slots == 0
+            or len(execute_model_req.seq_group_metadata_list) == 0
+            or disable_all_speculation,
```
Why did you need to modify this?

cadedaniel replied:
Got it. Can you add a comment?

cadedaniel commented on vllm/spec_decode/spec_decode_worker.py:
```diff
@@ -478,6 +484,19 @@ def _run_no_spec(self, execute_model_req: ExecuteModelRequest,
                     execute_model_req.seq_group_metadata_list, hidden_states)
 
+        if not skip_proposer:
+            # For prefill step in proposer, we run the model for N-1 tokens
+            # because Nth token will be processed in the first decode step. For
+            # N-1 tokens, the input should be 0:N-1 hidden states which should
+            # be concatanated with 1:N token (since output of scorer has to be
+            # the input for proposer). Therefore, we shift the hidden states to
+            # align n-1th hidden state with nth token.
+            if sampler_output.prefill_hidden_states is not None:
+                execute_model_req.previous_hidden_states = sampler_output\
+                    .prefill_hidden_states.roll(shifts=1, dims=0)
```
Thanks for adding a great comment.

Can this logic be pushed into the draft worker?

cadedaniel replied:
Ah got it. Nice.

Consider adding a comment explaining why this is here and moving this to a separate helper function (concern here is too much edge-case logic in the main flow makes it hard to read).

cadedaniel reviewed (commented):
Overall looks OK. Left some comments.

We can merge this after addressing the comments, but I am not confident in the correctness of the modeling. We should expect accuracy rates >=0.6 on sharegpt, not ~0.335. So we will need to fix that before we can recommend this for users.

abhigoyal1997 committed:
adding doctrings to EAGLE and Medusa models

abhigoyal1997 committed:
fixing hidden states handling in batch expansion

abhigoyal1997 committed:
making HiddenStates a dataclass and renaming last_non_bonus_hidden_states

abhigoyal1997 committed:
Merge branch 'hidden_states_fix' of github.fkinternal.com:abhinav-goyal/vllm into eagle

abhigoyal1997 committed:
reformatting

abhigoyal1997 commented on vllm/spec_decode/batch_expansion.py:
These changes were pulled from #7508

abhigoyal1997 commented on vllm/spec_decode/top1_proposer.py:
Part of #7508

abhigoyal1997 commented on vllm/spec_decode/util.py:
Part of #7508

abhigoyal1997 commented:
@cadedaniel Thanks for the review. I've addressed most of your comments. To close the remaining ones, I'll need your inputs on https://github.com/vllm-project/vllm/pull/6830/files#r1714096610 and https://github.com/vllm-project/vllm/pull/6830/files#r1714091225

cadedaniel commented:
LMK when it's ready for review

abhigoyal1997 committed:
adding acceptance rate test for large output length

abhigoyal1997 committed:
fixing hidden states manipulation for batch expansion

abhigoyal1997 committed:
Merge branch 'hidden_states_fix' of github.fkinternal.com:abhinav-goyal/vllm into eagle

abhigoyal1997 committed:
print acceptance rate in spec decode tests

abhigoyal1997 committed:
Updating HiddenStates to handle prefill step as well

abhigoyal1997 committed:
changing expected acceptance rate for test

abhigoyal1997 committed:
Merge branch 'vllm-project:main' into hidden_states_fix

abhigoyal1997 committed:
Merge branch 'vllm-project:main' into hidden_states_fix

abhigoyal1997 committed:
Merge branch 'hidden_states_fix' into eagle

abhigoyal1997 committed:
Adding explanation for trucated vocab and merging main

abhigoyal1997 committed:
formatting

abhigoyal1997 reviewed (commented):
@cadedaniel I've addressed the comments. PTAL, thanks!

cadedaniel commented:
Will review tomorrow !

abhigoyal1997 committed:
Merge branch 'main' of github.fkinternal.com:abhinav-goyal/vllm into eagle

abhigoyal1997 committed:
Fixing compatibility of `worker.multi_step_worker.MultiStepWorker` with `worker.worker_base.LocalOrDistributedWorkerBase`

abhigoyal1997 commented on vllm/sequence.py:
```diff
@@ -1176,40 +1180,87 @@ class HiddenStates(msgspec.Struct, array_like=True,
+    # all tokens, whereas for decode step, it use used for last accepted tokens.
     hidden_states: torch.Tensor
+    # The sequence group metadata list. Only needed for decode step.
+    seq_group_metadata_list: Optional[List[SequenceGroupMetadata]] = None
```
`seq_group_metadata_list` is not optional as it's not needed for prefill step.

abhigoyal1997 committed:
Merge branch 'main' into eagle

cadedaniel commented on tests/spec_decode/e2e/test_eagle_correctness.py:
```diff
@@ -0,0 +1,268 @@
+                                         max_output_len=output_len,
+                                         force_output_len=True)
+
+
```
I think we should also include a test that the accuracy meets the expectation, like what you added in https://github.com/vllm-project/vllm/pull/7508.

my worry is we have no e2e test for the correctness of the autoregressive hidden state handling. since rejection sampling guarantees distribution matches target model, we may not catch it until much later. what do you think?

abhigoyal1997 replied:
I agree that we should have E2E tests based on expected acceptance rates. If we have some reference numbers that we can use, we can add some test here as well.

As for testing autoregressive hidden state handling, that is already captured in #7508, no? Since MLPSpeculator also uses autoregressive hidden states in the same manner.

cadedaniel replied:
> As for testing autoregressive hidden state handling, that is already captured in https://github.com/vllm-project/vllm/pull/7508, no? Since MLPSpeculator also uses autoregressive hidden states in the same manner.

ah that's right! ok would be nice to have but let's not block!

cadedaniel commented on vllm/model_executor/models/medusa.py:
```diff
@@ -30,6 +30,19 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
 
 
 class Medusa(nn.Module):
+    """This class implements the Medusa draft model from the paper: https://arxiv.org/abs/2401.10774
+    Reference implementation: https://github.com/FasterDecoding/Medusa
+    
+    Differences from reference implementation:
+    1. Currently this only supports generating proposals from top-1 tokens.
+    2. We have an optional token_map which reduces draft vocab to most 
```
thanks for adding !

cadedaniel commented on vllm/sequence.py:
```diff
@@ -1176,40 +1180,87 @@ class HiddenStates(msgspec.Struct, array_like=True,
+                    .second_last_token_hidden_states[index]
             self._seq_ids = seq_ids
 
+    def expand_with_bonus_tokens(
```
(not blocking for this PR): these datastructures which require torch operations should not live in sequence // should go under `spec_decode`.

cadedaniel reviewed (commented):
sorry, need 1 more day to finish review. partial comments below.

abhigoyal1997 committed:
Merge branch 'main' into eagle

cadedaniel commented on vllm/worker/worker_base.py:
```diff
@@ -439,3 +458,23 @@ def execute_method(self, method, *args, **kwargs):
+def extract_previous_hidden_states(
+        data: Union[ExecuteModelRequest, Dict[str, torch.Tensor]]) -> \
+            Dict[str, torch.Tensor]:
+    """If data contains previous_hidden_states, extract it. This returns a dict
+    which can be used directly as additional kwargs in any following 
+    execute_model calls."""
+    output = {}
```
nit: mention this is used in Eagle

cadedaniel reviewed (approved):
LGTM! let's merge once tests are passing

abhigoyal1997 committed:
adding comment

abhigoyal1997 commented:
Thanks @cadedaniel for reviewing and approving. This is ready to merge!

cadedaniel commented:
do you have the steps for creating the checkpoint @abhigoyal1997  ?

jokmingwong commented:
```bash
python3 benchmark_latency.py --model Qwen/Qwen2-7B-Instruct --speculative-model yuhuili/EAGLE-Qwen2-7B-Instruct --num_speculative_tokens 4 --use-v2-block-manager ----batch-size 1 --input-len 1024--output-len 128 --max-model-len 2048
```
I used the command above to run [the latest Eagle speculative decoding PR](https://github.com/vllm-project/vllm/pull/6830) with Qwen2-7B. And the eagle model has been converted according to the conversion script in the [comments](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/eagle.py#L126-L131). After running, I found that the output generated by speculative decoding is inconsistent with that without speculative decoding. Upon examining the source code, I suspect that there may be an issue with the implementation about input_ids as the implementation of vLLM is inconsistent with the Eagle library implementation:
https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/eagle.py#L76-L99
https://github.com/SafeAILab/EAGLE/blob/main/eagle/modeling_eagle.py#L1273
Namely, when running the forward pass for the Eagle model, the first token of the original model should also be placed on the input_ids. Can you explain the reason for the difference in these two implementations?

abhigoyal1997 commented:
> ```shell
> python3 benchmarks/benchmark_latency.py --model Qwen/Qwen2-7B-Instruct --speculative-model yuhuili/EAGLE-Qwen2-7B-Instruct --num_speculative_tokens 4 --use-v2-block-manager --batch-size 1 --input-len 1024--output-len 128 --max-model-len 2048
> ```
> 
> I used the command above to run [the latest Eagle speculative decoding PR](https://github.com/vllm-project/vllm/pull/6830) with Qwen2-7B. And the eagle model has been converted according to the conversion script in the [comments](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/eagle.py#L126-L131). After running, I found that the output generated by speculative decoding is inconsistent with that without speculative decoding. Upon examining the source code, I suspect that there may be an issue with the implementation about input_ids as the implementation of vLLM is inconsistent with the Eagle library implementation: https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/eagle.py#L76-L99 https://github.com/SafeAILab/EAGLE/blob/main/eagle/modeling_eagle.py#L1273 Namely, when running the forward pass for the Eagle model, the first token of the original model should also be placed on the input_ids. Can you explain the reason for the difference in these two implementations?

Even in vLLM, the first token of the target model is present in input_ids. This is because first token is generated by target model in prefill step and EAGLE only starts generating token in the subsequent decode step. As for the masking in forward pass, that masks the first input token and not any token output by the target model. This didn't make any different in the outputs.

As for why you are seeing inconsistency, if you are using 16-bit precision, could it be related to this: https://github.com/vllm-project/vllm/pull/6830/files#r1714081377 ?

abhigoyal1997 commented:
> do you have the steps for creating the checkpoint @abhigoyal1997 ?

Is this helpful: https://github.com/vllm-project/vllm/blob/80c7b089b1189c5e2f40b3be250a118e9349a024/vllm/model_executor/models/eagle.py#L127 ?

jokmingwong commented:
> > ```shell
> > python3 benchmarks/benchmark_latency.py --model Qwen/Qwen2-7B-Instruct --speculative-model yuhuili/EAGLE-Qwen2-7B-Instruct --num_speculative_tokens 4 --use-v2-block-manager --batch-size 1 --input-len 1024--output-len 128 --max-model-len 2048
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I used the command above to run [the latest Eagle speculative decoding PR](https://github.com/vllm-project/vllm/pull/6830) with Qwen2-7B. And the eagle model has been converted according to the conversion script in the [comments](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/eagle.py#L126-L131). After running, I found that the output generated by speculative decoding is inconsistent with that without speculative decoding. Upon examining the source code, I suspect that there may be an issue with the implementation about input_ids as the implementation of vLLM is inconsistent with the Eagle library implementation: https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/eagle.py#L76-L99 https://github.com/SafeAILab/EAGLE/blob/main/eagle/modeling_eagle.py#L1273 Namely, when running the forward pass for the Eagle model, the first token of the original model should also be placed on the input_ids. Can you explain the reason for the difference in these two implementations?
> 
> Even in vLLM, the first token of the target model is present in input_ids. This is because the first token is generated by the target model in the prefill step which is then added to input_ids and EAGLE only starts generating tokens in the subsequent decode step. As for the masking in the forward pass, that masks the first input token and not any token output by the target model. This didn't make any difference in the outputs.
> 
> As for why you are seeing inconsistency, if you are using 16-bit precision, could it be related to this: [#4978 (comment)](https://github.com/vllm-project/vllm/pull/4978#discussion_r1669999328) ?

Sorry, I may not have expressed this issue clearly. I noticed that when using speculative decoding with Eagle in vLLM under the same prompt and same models, with top_k=1 and temperature=0.5, the output is inconsistent with the official Eagle implementation. I will provide my test cases later to help reproduce this issue.

Siegfried-qgf commented:
Great work. Do you have any plans to implement tree decoding? It seems that tree decoding will be very important to improve the results.

Sekri0 commented:
Do you have any plans to support scenarios where tp > 1? @abhigoyal1997

den-run-ai commented:
Is it possible to reproduce the ~2x-3x speedup reported in EAGLE 1/2 papers with this PR in vLLM?

xiongqisong commented:
> ```shell
> python3 benchmarks/benchmark_latency.py --model Qwen/Qwen2-7B-Instruct --speculative-model yuhuili/EAGLE-Qwen2-7B-Instruct --num_speculative_tokens 4 --use-v2-block-manager --batch-size 1 --input-len 1024--output-len 128 --max-model-len 2048
> ```
> 
> I used the command above to run [the latest Eagle speculative decoding PR](https://github.com/vllm-project/vllm/pull/6830) with Qwen2-7B. And the eagle model has been converted according to the conversion script in the [comments](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/eagle.py#L126-L131). After running, I found that the output generated by speculative decoding is inconsistent with that without speculative decoding. Upon examining the source code, I suspect that there may be an issue with the implementation about input_ids as the implementation of vLLM is inconsistent with the Eagle library implementation: https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/eagle.py#L76-L99 https://github.com/SafeAILab/EAGLE/blob/main/eagle/modeling_eagle.py#L1273 Namely, when running the forward pass for the Eagle model, the first token of the original model should also be placed on the input_ids. Can you explain the reason for the difference in these two implementations?

I try to use EAGLE on Llama but failed, i want to know how to use EAGLE on vLLM, it's so hard to use with no demo

sroy745 commented:
Hi @xiongqisong, which checkpoint are you using as the draft model? Is it one of the checkpoints available [here](https://huggingface.co/yuhuili)  ? If so it will not work since the checkpoint needed in vLLM is a bit different from what is available at https://huggingface.co/yuhuili. You need to convert the checkpoint available in https://huggingface.co/yuhuili using the script [here](https://gist.github.com/abhigoyal1997/1e7a4109ccb7704fbc67f625e86b2d6d) and use the converted checkpoint as the draft model in the vllm.

Please let us know if this works for you or not. I think @LiuXiaoxuanPKU recently used the script to convert the checkpoint for yuhuili/EAGLE-LLaMA3-Instruct-70B into the vLLM compatible checkpoint and it worked for her.

cc: @LiuXiaoxuanPKU

xiongqisong commented:
> Hi @xiongqisong, which checkpoint are you using as the draft model? Is it one of the checkpoints available [here](https://huggingface.co/yuhuili) ? If so it will not work since the checkpoint needed in vLLM is a bit different from what is available at https://huggingface.co/yuhuili. You need to convert the checkpoint available in https://huggingface.co/yuhuili using the script [here](https://gist.github.com/abhigoyal1997/1e7a4109ccb7704fbc67f625e86b2d6d) and use the converted checkpoint as the draft model in the vllm.
> 
> Please let us know if this works for you or not. I think @LiuXiaoxuanPKU recently used the script to convert the checkpoint for yuhuili/EAGLE-LLaMA3-Instruct-70B into the vLLM compatible checkpoint and it worked for her.
> 
> I will add a section on how to use Eagle to the sd documentation [here](https://docs.vllm.ai/en/latest/usage/spec_decode.html) shortly
> 
> cc: @LiuXiaoxuanPKU

Thanks for reply, i already use the script to convert EAGLE model weight to vllm format weight, but vLLM can't run EAGLE normally, i share the clue in #11126 , hope you have time to help me @sroy745 ~

Kamichanw commented:
Can someone help me with the propose of `_expand_execute_model_request` here?
https://github.com/vllm-project/vllm/blob/289b5191d5bd16c7b2e08cb19434f9f188393ac5/vllm/spec_decode/multi_step_worker.py#L86-L91
I can't understand why it copies a new model seq without bonus token and sends to `execute_model`. Finally, at here
https://github.com/vllm-project/vllm/blob/289b5191d5bd16c7b2e08cb19434f9f188393ac5/vllm/spec_decode/multi_step_worker.py#L126-L127
this function select the original seq. What's the difference with or without the model seq without bonus token created before?
@abhigoyal1997
